providers/implementations/ciphers/ciphercommon_block.c

```diff
@@ -120,18 +131,19 @@ int ossl_cipher_unpadblock(unsigned char *buf, size_t *buflen, size_t blocksize)
      * Otherwise it provides a padding oracle.
      */
     pad = buf[blocksize - 1];
-    if (pad == 0 || pad > blocksize) {
-        ERR_raise(ERR_LIB_PROV, PROV_R_BAD_DECRYPT);
-        return 0;
+    pad_ok = ~(value_barrier_s(constant_time_is_zero_s(pad)) |
+                value_barrier_s(constant_time_lt_s(blocksize, pad)));
+    gp = constant_time_select_s(pad_ok, CONSTTIME_TRUE, CONSTTIME_FALSE);
+
+    for (i = 0; i < blocksize; i++) {
+        i_pad_lt = constant_time_lt(i, pad);
+        len = constant_time_select_s(i_pad_lt, len - 1, len);
+        buf_pad_eq = constant_time_eq_s(buf[len], pad);
+        gp = constant_time_select_s(i_pad_lt, gp & buf_pad_eq, gp);
     }
-    for (i = 0; i < pad; i++) {
-        if (buf[--len] != pad) {
-            ERR_raise(ERR_LIB_PROV, PROV_R_BAD_DECRYPT);
```

Comment:
we have done a similar surgery in the rsa_pk1.c,
at that time we considered removing the ERR_raise,
but it was arguably an API change,
there we have used a constant time error function:
```C
    RSAerr(RSA_F_RSA_PADDING_CHECK_PKCS1_TYPE_2, RSA_R_PKCS_DECODING_ERROR);
    err_clear_last_constant_time(1 & good);
```
that is raise the error unconditionally, and conditionally clear it again in constant time.

Reply:
This is exactly what was needed. Thanks for the pointer!

Reply:
If you look at rsa_pk1.c, there is also a comment explicitly stating the following:
```c
/*
  * This trick doesn't work in the FIPS provider because libcrypto manages
  * the error stack. Instead we opt not to put an error on the stack at all
  * in case of padding failure in the FIPS provider.
*/
```
I don't think we can have a constant-time `ossl_cipher_unpadblock` without an API change in this case.

Reply:
The only solution seems to be exposing `err_clear_last_constant_time` to FIPS code but I am not sure this is possible.

Reply:
The FIPS code isn't allowed to link any symbols from libcrypto.
It is allowed to have a local copy of what it needs.

Reply:
...  or have it map to the core upcall...  which is already done (see the change in fipsprov.c)